File: sdk/src/profiles/wire/encoder.js
// weavepack-wire — encoder (schemaless snapshots + delta chains).
//
// Profile isolation: imports only from ./types.js. No JSON/tensor profile code.
//
// Binary format summary:
//   byte 0: FLAG_SCHEMALESS (0x00) | FLAG_DELTA (0x01) | FLAG_SCHEMAFUL (0x02)
//
// Schemaless snapshot body:
//   LEB128(field_count)
//   for each field (ascending field_number order):
//     LEB128(field_number)
//     1 byte type_tag  (scalarTag or containerTag)
//     value bytes
//
// Delta chain body:
//   LEB128(op_count)
//   for each op: 1 byte op_code + path bytes + value bytes

import {
  VTYPE, CTYPE, OP, PC,
  FLAG_SCHEMALESS, FLAG_DELTA,
  scalarTag, containerTag, isContainer,
  TAG_MESSAGE, TAG_REPEATED, TAG_MAP, TAG_ONEOF,
  MAX_PAYLOAD_BYTES,
} from "./types.js"

const _enc = new TextEncoder()

// ── ByteWriter ─────────────────────────────────────────────────────────────

class ByteWriter {
  constructor() {
    this._buf = []
  }

  writeByte(b) {
    this._buf.push(b & 0xFF)
  }

  writeBytes(src) {
    for (let i = 0; i < src.length; i++) this._buf.push(src[i] & 0xFF)
  }

  writeLEB128(v) {
    if (typeof v === "bigint") {
      while (v >= 128n) {
        this._buf.push(Number(v & 0x7Fn) | 0x80)
        v >>= 7n
      }
      this._buf.push(Number(v))
    } else {
      v = Math.floor(v) >>> 0
      while (v >= 128) {
        this._buf.push((v & 0x7F) | 0x80)
        v = (v >>> 7)
      }
      this._buf.push(v)
    }
  }

  toBytes() { return new Uint8Array(this._buf) }
}

// ── Scalar encoding ────────────────────────────────────────────────────────

function writeScalar(w, vtype, value) {
  switch (vtype) {
    case VTYPE.BOOL:
      w.writeByte(value ? 1 : 0)
      break
    case VTYPE.INT32: {
      // Two's-complement bit pattern as unsigned LEB128.
      w.writeLEB128((value | 0) >>> 0)
      break
    }
    case VTYPE.INT64: {
      // Caller provides BigInt.
      const bv = BigInt(value)
      const u = bv < 0n ? bv + (1n << 64n) : bv
      w.writeLEB128(u)
      break
    }
    case VTYPE.UINT32:
      w.writeLEB128(value >>> 0)
      break
    case VTYPE.UINT64:
      w.writeLEB128(BigInt(value))
      break
    case VTYPE.SINT32: {
      const v = value | 0
      // Zigzag: small negatives → small positives.
      const z = ((v << 1) ^ (v >> 31)) >>> 0
      w.writeLEB128(z)
      break
    }
    case VTYPE.SINT64: {
      const bv = BigInt(value)
      const z = ((bv << 1n) ^ (bv >> 63n)) & 0xFFFFFFFFFFFFFFFFn
      w.writeLEB128(z)
      break
    }
    case VTYPE.FLOAT32: {
      const tmp = new ArrayBuffer(4)
      new DataView(tmp).setFloat32(0, value, true)
      w.writeBytes(new Uint8Array(tmp))
      break
    }
    case VTYPE.FLOAT64: {
      const tmp = new ArrayBuffer(8)
      new DataView(tmp).setFloat64(0, value, true)
      w.writeBytes(new Uint8Array(tmp))
      break
    }
    case VTYPE.STRING: {
      const utf8 = _enc.encode(value)
      if (utf8.length > MAX_PAYLOAD_BYTES) throw new Error("string exceeds 256 MiB limit")
      w.writeLEB128(utf8.length)
      w.writeBytes(utf8)
      break
    }
    case VTYPE.BYTES: {
      const src = value instanceof Uint8Array ? value : new Uint8Array(value)
      if (src.length > MAX_PAYLOAD_BYTES) throw new Error("bytes exceeds 256 MiB limit")
      w.writeLEB128(src.length)
      w.writeBytes(src)
      break
    }
    case VTYPE.ENUM: {
      // Stored as int32 two's-complement unsigned LEB128.
      w.writeLEB128((value | 0) >>> 0)
      break
    }
    default:
      throw new Error(`unknown vtype ${vtype}`)
  }
}

// ── Path encoding ──────────────────────────────────────────────────────────
//
// Path: sequence of components, each prefixed by 1 byte component-type.
//   0 (PC.FIELD): LEB128 field_number
//   1 (PC.MAP):   1 byte key_type (0=string,1=uint32) + key
//   2 (PC.INDEX): LEB128 index
//   3 (PC.END):   (no payload) — terminates the path
//
// A path of [] (root message) is encoded as just a single END byte.

function writePath(w, path) {
  for (const comp of path) {
    if (comp.field !== undefined) {
      w.writeByte(PC.FIELD)
      w.writeLEB128(comp.field >>> 0)
    } else if (comp.map !== undefined) {
      w.writeByte(PC.MAP)
      if (typeof comp.map === "string") {
        w.writeByte(0)  // string key
        const utf8 = _enc.encode(comp.map)
        w.writeLEB128(utf8.length)
        w.writeBytes(utf8)
      } else {
        w.writeByte(1)  // uint32 key
        w.writeLEB128(comp.map >>> 0)
      }
    } else if (comp.index !== undefined) {
      w.writeByte(PC.INDEX)
      w.writeLEB128(comp.index >>> 0)
    }
  }
  w.writeByte(PC.END)
}

// ── Field/message encoding ─────────────────────────────────────────────────
//
// A field object is one of:
//   {num, vtype, value}                                          — scalar
//   {num, message: Field[]}                                      — nested message
//   {num, repeated: {elemType, values}}                          — repeated scalar
//   {num, map: {keyType, valueType, entries}}                    — map
//     entries: Array of [key, value] pairs
//     keyType: 'string' | 'uint32'
//     valueType: vtype (scalar only in v0.1)
//   {num, oneof: {activeField, valueType, value}}                — oneof

function writeFieldBody(w, field) {
  if (field.message !== undefined) {
    w.writeByte(TAG_MESSAGE)
    writeMessageBody(w, field.message)
  } else if (field.repeated !== undefined) {
    w.writeByte(TAG_REPEATED)
    const { elemType, values } = field.repeated
    w.writeByte(scalarTag(elemType))
    w.writeLEB128(values.length)
    for (const v of values) writeScalar(w, elemType, v)
  } else if (field.map !== undefined) {
    w.writeByte(TAG_MAP)
    const { keyType, valueType, entries } = field.map
    w.writeByte(keyType === "string" ? 0 : 1)
    w.writeByte(scalarTag(valueType))
    w.writeLEB128(entries.length)
    for (const [k, v] of entries) {
      if (keyType === "string") {
        const utf8 = _enc.encode(k)
        w.writeLEB128(utf8.length)
        w.writeBytes(utf8)
      } else {
        w.writeLEB128(k >>> 0)
      }
      writeScalar(w, valueType, v)
    }
  } else if (field.oneof !== undefined) {
    w.writeByte(TAG_ONEOF)
    const { activeField, valueType, value } = field.oneof
    w.writeLEB128(activeField >>> 0)
    w.writeByte(scalarTag(valueType))
    writeScalar(w, valueType, value)
  } else {
    // Scalar
    const { vtype, value } = field
    w.writeByte(scalarTag(vtype))
    writeScalar(w, vtype, value)
  }
}

function writeMessageBody(w, fields) {
  // Sort by field number ascending (canonical form).
  const sorted = [...fields].sort((a, b) => a.num - b.num)
  w.writeLEB128(sorted.length)
  for (const f of sorted) {
    w.writeLEB128(f.num >>> 0)
    writeFieldBody(w, f)
  }
}

// ── Public API: snapshots ──────────────────────────────────────────────────

export function encodeDocument(fields) {
  const w = new ByteWriter()
  w.writeByte(FLAG_SCHEMALESS)
  writeMessageBody(w, fields)
  return w.toBytes()
}

// ── Delta op encoding ──────────────────────────────────────────────────────

function writeOp(w, op) {
  w.writeByte(op.op)
  writePath(w, op.path ?? [])

  switch (op.op) {
    case OP.FIELD_SET:
      // path + type_tag + value
      writeFieldBody(w, { num: 0, ...op.value })
      break
    case OP.FIELD_DELETE:
      // path only (no value)
      break
    case OP.MESSAGE_REPLACE:
      // path + message body
      writeMessageBody(w, op.message)
      break
    case OP.REPEATED_APPEND: {
      const { elemType, values } = op.elements
      w.writeByte(scalarTag(elemType))
      w.writeLEB128(values.length)
      for (const v of values) writeScalar(w, elemType, v)
      break
    }
    case OP.REPEATED_SPLICE: {
      const { index, deleteCount, elemType, insertValues } = op
      w.writeLEB128(index >>> 0)
      w.writeLEB128(deleteCount >>> 0)
      w.writeByte(scalarTag(elemType))
      w.writeLEB128(insertValues.length)
      for (const v of insertValues) writeScalar(w, elemType, v)
      break
    }
    case OP.MAP_SET: {
      const { keyType, key, valueType, value } = op
      w.writeByte(keyType === "string" ? 0 : 1)
      if (keyType === "string") {
        const utf8 = _enc.encode(key)
        w.writeLEB128(utf8.length)
        w.writeBytes(utf8)
      } else {
        w.writeLEB128(key >>> 0)
      }
      w.writeByte(scalarTag(valueType))
      writeScalar(w, valueType, value)
      break
    }
    case OP.MAP_DELETE: {
      const { keyType, key } = op
      w.writeByte(keyType === "string" ? 0 : 1)
      if (keyType === "string") {
        const utf8 = _enc.encode(key)
        w.writeLEB128(utf8.length)
        w.writeBytes(utf8)
      } else {
        w.writeLEB128(key >>> 0)
      }
      break
    }
    case OP.ONEOF_SWITCH: {
      const { activeField, valueType, value } = op
      w.writeLEB128(activeField >>> 0)
      w.writeByte(scalarTag(valueType))
      writeScalar(w, valueType, value)
      break
    }
    default:
      throw new Error(`unknown op ${op.op}`)
  }
}

export function encodeChain(ops) {
  const w = new ByteWriter()
  w.writeByte(FLAG_DELTA)
  w.writeLEB128(ops.length)
  for (const op of ops) writeOp(w, op)
  return w.toBytes()
}

File: impl/python/weavepack_wire/encoder.py
"""weavepack-wire — pure-Python encoder (schemaless snapshots + delta chains).

Profile isolation: imports only from .types. No JSON/tensor profile code.
"""

import struct
from .types import (
    VTYPE, CTYPE, OP, PC,
    FLAG_SCHEMALESS, FLAG_DELTA,
    scalar_tag, container_tag, is_container,
    TAG_MESSAGE, TAG_REPEATED, TAG_MAP, TAG_ONEOF,
    MAX_PAYLOAD_BYTES,
)


class _ByteWriter:
    def __init__(self):
        self._buf = bytearray()

    def write_byte(self, b: int):
        self._buf.append(b & 0xFF)

    def write_bytes(self, src: bytes | bytearray | memoryview):
        self._buf.extend(src)

    def write_leb128(self, v: int):
        if isinstance(v, int) and v < 0:
            raise ValueError(f"write_leb128: negative value {v}")
        v = int(v)
        while v >= 128:
            self._buf.append((v & 0x7F) | 0x80)
            v >>= 7
        self._buf.append(v)

    def to_bytes(self) -> bytes:
        return bytes(self._buf)


def _write_scalar(w: _ByteWriter, vtype: int, value):
    if vtype == VTYPE.BOOL:
        w.write_byte(1 if value else 0)
    elif vtype == VTYPE.INT32:
        v = int(value) & 0xFFFFFFFF
        w.write_leb128(v)
    elif vtype == VTYPE.INT64:
        v = int(value)
        if v < 0:
            v = v + (1 << 64)
        w.write_leb128(v)
    elif vtype == VTYPE.UINT32:
        w.write_leb128(int(value) & 0xFFFFFFFF)
    elif vtype == VTYPE.UINT64:
        w.write_leb128(int(value))
    elif vtype == VTYPE.SINT32:
        v = int(value) & 0xFFFFFFFF if int(value) >= 0 else int(value)
        # Zigzag: small negatives → small positives.
        v = int(value)
        z = ((v << 1) ^ (v >> 31)) & 0xFFFFFFFF
        w.write_leb128(z)
    elif vtype == VTYPE.SINT64:
        v = int(value)
        z = ((v << 1) ^ (v >> 63)) & 0xFFFFFFFFFFFFFFFF
        w.write_leb128(z)
    elif vtype == VTYPE.FLOAT32:
        w.write_bytes(struct.pack("<f", float(value)))
    elif vtype == VTYPE.FLOAT64:
        w.write_bytes(struct.pack("<d", float(value)))
    elif vtype == VTYPE.STRING:
        utf8 = str(value).encode("utf-8")
        if len(utf8) > MAX_PAYLOAD_BYTES:
            raise ValueError("string exceeds 256 MiB limit")
        w.write_leb128(len(utf8))
        w.write_bytes(utf8)
    elif vtype == VTYPE.BYTES:
        if isinstance(value, (bytes, bytearray)):
            src = bytes(value)
        elif isinstance(value, list):
            src = bytes(value)
        else:
            src = bytes(value)
        if len(src) > MAX_PAYLOAD_BYTES:
            raise ValueError("bytes exceeds 256 MiB limit")
        w.write_leb128(len(src))
        w.write_bytes(src)
    elif vtype == VTYPE.ENUM:
        v = int(value) & 0xFFFFFFFF
        w.write_leb128(v)
    else:
        raise ValueError(f"unknown vtype {vtype}")


def _write_path(w: _ByteWriter, path: list):
    for comp in path:
        if "field" in comp:
            w.write_byte(PC.FIELD)
            w.write_leb128(int(comp["field"]) & 0xFFFFFFFF)
        elif "map" in comp:
            w.write_byte(PC.MAP)
            key = comp["map"]
            if isinstance(key, str):
                w.write_byte(0)  # string key
                utf8 = key.encode("utf-8")
                w.write_leb128(len(utf8))
                w.write_bytes(utf8)
            else:
                w.write_byte(1)  # uint32 key
                w.write_leb128(int(key) & 0xFFFFFFFF)
        elif "index" in comp:
            w.write_byte(PC.INDEX)
            w.write_leb128(int(comp["index"]) & 0xFFFFFFFF)
    w.write_byte(PC.END)


def _write_field_body(w: _ByteWriter, field: dict):
    if "message" in field:
        w.write_byte(TAG_MESSAGE)
        _write_message_body(w, field["message"])
    elif "repeated" in field:
        w.write_byte(TAG_REPEATED)
        r = field["repeated"]
        elem_type = r["elemType"]
        values = r["values"]
        w.write_byte(scalar_tag(elem_type))
        w.write_leb128(len(values))
        for v in values:
            _write_scalar(w, elem_type, v)
    elif "map" in field:
        w.write_byte(TAG_MAP)
        m = field["map"]
        key_type = m["keyType"]
        value_type = m["valueType"]
        entries = m["entries"]
        w.write_byte(0 if key_type == "string" else 1)
        w.write_byte(scalar_tag(value_type))
        w.write_leb128(len(entries))
        for k, v in entries:
            if key_type == "string":
                utf8 = str(k).encode("utf-8")
                w.write_leb128(len(utf8))
                w.write_bytes(utf8)
            else:
                w.write_leb128(int(k) & 0xFFFFFFFF)
            _write_scalar(w, value_type, v)
    elif "oneof" in field:
        w.write_byte(TAG_ONEOF)
        o = field["oneof"]
        w.write_leb128(int(o["activeField"]) & 0xFFFFFFFF)
        w.write_byte(scalar_tag(o["valueType"]))
        _write_scalar(w, o["valueType"], o["value"])
    else:
        vtype = field["vtype"]
        value = field["value"]
        w.write_byte(scalar_tag(vtype))
        _write_scalar(w, vtype, value)


def _write_message_body(w: _ByteWriter, fields: list):
    sorted_fields = sorted(fields, key=lambda f: f["num"])
    w.write_leb128(len(sorted_fields))
    for f in sorted_fields:
        w.write_leb128(int(f["num"]) & 0xFFFFFFFF)
        _write_field_body(w, f)


def encode_document(fields: list) -> bytes:
    """Encode a schemaless snapshot from a list of field dicts."""
    w = _ByteWriter()
    w.write_byte(FLAG_SCHEMALESS)
    _write_message_body(w, fields)
    return w.to_bytes()


def _write_op(w: _ByteWriter, op: dict):
    w.write_byte(op["op"])
    _write_path(w, op.get("path") or [])

    op_code = op["op"]
    if op_code == OP.FIELD_SET:
        _write_field_body(w, {"num": 0, **op["value"]})
    elif op_code == OP.FIELD_DELETE:
        pass  # path only
    elif op_code == OP.MESSAGE_REPLACE:
        _write_message_body(w, op["message"])
    elif op_code == OP.REPEATED_APPEND:
        elems = op["elements"]
        elem_type = elems["elemType"]
        values = elems["values"]
        w.write_byte(scalar_tag(elem_type))
        w.write_leb128(len(values))
        for v in values:
            _write_scalar(w, elem_type, v)
    elif op_code == OP.REPEATED_SPLICE:
        w.write_leb128(int(op["index"]) & 0xFFFFFFFF)
        w.write_leb128(int(op["deleteCount"]) & 0xFFFFFFFF)
        elem_type = op["elemType"]
        insert_values = op["insertValues"]
        w.write_byte(scalar_tag(elem_type))
        w.write_leb128(len(insert_values))
        for v in insert_values:
            _write_scalar(w, elem_type, v)
    elif op_code == OP.MAP_SET:
        key_type = op["keyType"]
        key = op["key"]
        value_type = op["valueType"]
        value = op["value"]
        w.write_byte(0 if key_type == "string" else 1)
        if key_type == "string":
            utf8 = str(key).encode("utf-8")
            w.write_leb128(len(utf8))
            w.write_bytes(utf8)
        else:
            w.write_leb128(int(key) & 0xFFFFFFFF)
        w.write_byte(scalar_tag(value_type))
        _write_scalar(w, value_type, value)
    elif op_code == OP.MAP_DELETE:
        key_type = op["keyType"]
        key = op["key"]
        w.write_byte(0 if key_type == "string" else 1)
        if key_type == "string":
            utf8 = str(key).encode("utf-8")
            w.write_leb128(len(utf8))
            w.write_bytes(utf8)
        else:
            w.write_leb128(int(key) & 0xFFFFFFFF)
    elif op_code == OP.ONEOF_SWITCH:
        w.write_leb128(int(op["activeField"]) & 0xFFFFFFFF)
        value_type = op["valueType"]
        w.write_byte(scalar_tag(value_type))
        _write_scalar(w, value_type, op["value"])
    else:
        raise ValueError(f"unknown op {op_code}")


def encode_chain(ops: list) -> bytes:
    """Encode a delta chain from a list of op dicts."""
    w = _ByteWriter()
    w.write_byte(FLAG_DELTA)
    w.write_leb128(len(ops))
    for op in ops:
        _write_op(w, op)
    return w.to_bytes()
